specs/FileTypePolicy.md

```diff
@@ -0,0 +1,222 @@
+File Type Policy API
+===
+
+# Background
+When saving a file with original SaveFilePicker, a security alert might be
+prompted, because the browser applies the [file type policies](https://learn.microsoft.com/en-us/deployedge/microsoft-edge-security-downloads-interruptions#file-types-requiring-a-gesture)
+to protect end users. However, in a WebView2 build App, when end users try
+to save a file with a certain file extension, they usually can trust the
+host App, domain and file extension. So, we provide the App developers the
+File Type Policy API to manage the file type policies dynamically.
+
+We'd appreciate your feedback.
+
+# Description
+
+We proposed the `CoreWebView2.SaveFileSecurityCheckStarting` event. You can register
+this event to get the file path, file extension and URI source information,
+when end users try to save a file from your App. Then you can apply your own
+rules to allow save the file with, or without a default warning dialog;
+to cancel the saving; and even to create your own UI to manage runtime 
+file type policies.
+
+# Examples
+## Win32 C++ 
+This example shows suppressing file type policy, security dialog, and 
+allow to save the file directly. It also blocks saving the exe file.
+The sample code will register the event with custom rules.
+```c++
+bool ScenarioFileTypePolicyStaging::AddCustomFileTypePolicies()
+{
+    if (!m_webView2Staging25)
+        return false;
+    // Register a handler for the `SaveFileSecurityCheckStarting` event.
+    m_webView2Staging25->add_SaveFileSecurityCheckStarting(
+        Callback<ICoreWebView2StagingSaveFileSecurityCheckStartingEventHandler>(
+            [this](
+                ICoreWebView2* sender,
+                ICoreWebView2StagingSaveFileSecurityCheckStartingEventArgs* args) -> HRESULT
+            {
+                // Get the file extension for file to be saved.
+                // And create your own rules of file type policy.
+                LPWSTR extension;
+                CHECK_FAILURE(args->get_FileExtension(&extension));
+                if (lstrcmpW(extension, L"eml") == 0)
+                {
+                    // Set the `SuppressDefaultPolicy` property to skip the
+                    // default file type policy checking and a possible security
+                    // alert dialog for "eml" file. This will consent to
+                    // save the file.
+                    CHECK_FAILURE(args->put_SuppressDefaultPolicy(TRUE));
+                }
+                if (lstrcmpW(extension, L"exe") == 0)
```

Comment:
We shouldn't use exe as our example, because that implies that the sample is showing security best practices. (Which it's not. This code lets `.com` slip through, for example.) Instead, maybe we block some other file type like `.iso`.

Reply:
Change eml check to also check that the origin is from a trusted source (via a helper method that you don't define) and add an else clause that denies everything else with cancel and shows dialog telling end user that.